source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+You can manually trigger the start of the virtual machine by using
+```shell-session
+>>> machine.start()
+```
+for a specific node.
+
+or
+```shell-session
+>>> start_all()
+```
+for all specified nodes.
+
+You can enter a interactive shell on the virtual machine using:
+```shell-session
+>>> machine.shell_interact()
+```
+
+and run commandline commands like:
+```shell-session
+$ uname -a
+Linux server 5.10.37 #1-NixOS SMP Fri May 14 07:50:46 UTC 2021 x86_64 GNU/Linux
+```
+
+
+The nixos manual contains a [reference for commands](https://nixos.org/manual/nixos/stable/index.html#ssec-machine-objects) to design tests.
+
+
+### rerun successful tests
+
+Because test results are stored in the nix-store a successful test is cached, this means that nix will not run the test a second time as long as the test setup (configuration or test script) is not changed.
+To run a test again one needs to remove the result.
+If you would try to delete the result using the symbolic link you will get the following error:
+```shell-session
+$ nix-store --delete ./result
+finding garbage collector roots...
+0 store paths deleted, 0.00 MiB freed
+error: Cannot delete path '/nix/store/4klj06bsilkqkn6h2sia8dcsi72wbcfl-vm-test-run-unnamed' since it is still alive. To find out why, use: nix-store --query --roots
+```
+
+you need to remove the symbolic link and only than you can remove the cached result:
+```shell-session
+$ rm ./result
+```
+
+now you are able to remove the cached result in the nix store directly
+```shell-session
+$ nix-store --delete /nix/store/4klj06bsilkqkn6h2sia8dcsi72wbcfl-vm-test-run-unnamed
+```
+
+### wayland application example (todo remove home-manager)
+
+The configuration we are using is starting the gnome desktop manager using wayland.
+To test if a wayland application is working is more complicated because we need to automate the login into gnome and automated startup of the application. Additionally we need to enable access to gnome dbus interface. To do this we need to modify the configuration
+
+ the automated start of the application including automated login to gnome/wayland
+
+
+In the machine configuration we need to enable autologin for the user alice.
+
+```nix
+      services.xserver.displayManager.autoLogin.enable = true;
+      services.xserver.displayManager.autoLogin.user = "alice";
+```
+
+To simplify our script we pin the uid of the user to 1000.
+```nix
+        uid = 1000;
+```
+
+We specify a service that auto start firefox after login, which is easier than doing this in the test script.
+
+```nix
+      environment.systemPackages = [
+        (pkgs.makeAutostartItem {
+          name = "firefox";
+          package = pkgs.firefox;
+        })
+      ];
+```
+<!-- check again if this is best practice -->
+
+
+Because gnome doesn't allow the evaluation of javascript to get information about open windows we need to override the gnome-shell startup service to start gnome-shell in unsafe mode:
```

Comment:
Use original capitalization of software project names.
```suggestion
Since GNOME doesn't allow the evaluation of JavaScript to get information about open windows, we need to override the GNOME Shell startup service to start `gnome-shell` in unsafe mode:
```